cirq-core/cirq/contrib/hacks/disable_validation.py

```diff
@@ -0,0 +1,42 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tools for disabling validation in circuit construction."""
+
+import contextlib
+
+
+@contextlib.contextmanager
+def disable_op_validation(*, accept_debug_responsibility: bool = False):
+    if not accept_debug_responsibility:
+        raise ValueError(
+            "WARNING! Using disable_op_validation with invalid ops can cause "
+            "mysterious and terrible things to happen. cirq-maintainers will "
+            "not help you debug beyond this point!\n"
+            "If you still wish to continue, call this method with "
+            "accept_debug_responsibility=True."
+        )
+
+    from cirq.ops import raw_types
+
+    temp = raw_types._validate_qid_shape
+    raw_types._validate_qid_shape = lambda *args: None
```

Comment:
We might consider using [`contextvars`](https://docs.python.org/3/library/contextvars.html) instead:
```python
# raw_types.py
validate_qids: contextvars.ContextVar[bool] = contextvars.ContextVar('validate_qids', default=True)

def _validate_qid_shape(...) -> None:
    if validate_qids.get():
        return
    ...

# here
def disable_op_validation(...):
    ...
    token = raw_types.validate_qids.set(False)
    try:
        yield
    finally:
        raw_types.validate_qids.reset(token)
```
This would be more scoped than modifying global state, which could otherwise cause hard-to-debug issues with threads or asynchrony.

Reply:
Done. My only misgiving with this is that it exposes the hack in `raw_types.py`, which could lead to users applying it in ill-advised ways. Left a comment to mitigate this somewhat.

Reply:
After thinking about it, I think I agree with you that it's better to keep the hack itself contained and not modify `raw_types.py`. If this turns out to be useful and we can't otherwise optimize the qid validation then we could migrate it into raw_types in the future. So feel free to ignore my suggestions and revert to your original version.